source/grid/grid_tools.cc

```diff
@@ -5962,6 +5962,39 @@ namespace GridTools
 
   namespace internal
   {
+    std::tuple<std::vector<unsigned int>,
+               std::vector<unsigned int>,
+               std::vector<unsigned int>>
+    convert_guessed_ranks_and_indices_to_crs(
+      std::vector<std::pair<unsigned int, unsigned int>> &ranks_and_indices)
+    {
+      std::sort(ranks_and_indices.begin(), ranks_and_indices.end());
+
+      std::vector<unsigned int> ranks;
+      std::vector<unsigned int> ptr;
+      std::vector<unsigned int> indices;
+
+      unsigned int dummy_rank = numbers::invalid_unsigned_int;
+
+      for (const auto &i : ranks_and_indices)
+        {
+          if (dummy_rank != i.first)
+            {
+              dummy_rank = i.first;
+              ranks.push_back(dummy_rank);
+              ptr.push_back(indices.size());
```

Comment:
since `ranks` and `ptr` are the same size (except for the last element to do the crs lookup) and used together, have you thought about putting them into the same field, `std::vector<std::pair<unsigned int, unsigned int>>`? Not sure which one to prefer, but I typically prefer the latter, see e.g. https://github.com/dealii/dealii/blob/f1b6f2406dd091ea5aa80a55d1281fd99c098a5f/include/deal.II/base/partitioner.h#L714-L718